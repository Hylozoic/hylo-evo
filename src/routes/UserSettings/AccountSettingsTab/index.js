import AccountSettingsTab from './AccountSettingsTab'

export default AccountSettingsTab
