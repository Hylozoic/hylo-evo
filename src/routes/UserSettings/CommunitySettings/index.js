import CommunitySettings from './CommunitySettings'

export default CommunitySettings
