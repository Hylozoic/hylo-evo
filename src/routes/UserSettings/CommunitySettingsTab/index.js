import CommunitySettingsTab from './CommunitySettingsTab'

export default CommunitySettings
