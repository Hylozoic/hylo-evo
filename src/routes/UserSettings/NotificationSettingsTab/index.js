import NotificationSettingsTab from './NotificationSettingsTab'

export default NotificationSettingsTab
