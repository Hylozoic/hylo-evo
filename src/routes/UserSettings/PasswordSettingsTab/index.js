import PasswordSettingsTab from './PasswordSettingsTab'

export default PasswordSettingsTab
