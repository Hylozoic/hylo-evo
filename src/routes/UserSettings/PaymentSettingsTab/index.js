import PaymentSettingsTab from './PaymentSettingsTab'

export default PaymentSettingsTab
