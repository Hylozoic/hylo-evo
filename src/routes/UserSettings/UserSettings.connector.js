import { connect } from 'react-redux'
import getMe from 'store/selectors/getMe'
import {
  updateUserSettings, leaveCommunity, unlinkAccount,
  updateMembershipSettings, updateAllMemberships, registerStripeAccount
} from './UserSettings.store'
import { setConfirmBeforeClose } from '../FullPageModal/FullPageModal.store'
import { loginWithService } from 'routes/NonAuthLayout/Login/Login.store'
import { createSelector as ormCreateSelector } from 'redux-orm'
import { createSelector } from 'reselect'
import orm from 'store/models'
import fetchForCurrentUser from 'store/actions/fetchForCurrentUser'
import unBlockUser from 'store/actions/unBlockUser'
import getBlockedUsers from 'store/selectors/getBlockedUsers'
import { FETCH_FOR_CURRENT_USER } from 'store/constants'
import { get, every, includes } from 'lodash/fp'
import getQuerystringParam from 'store/selectors/getQuerystringParam'

// this selector assumes that all Memberships belong to the currentUser
// using this instead of currentUser.memberships to avoid a memoization issue
// see https://github.com/tommikaikkonen/redux-orm/issues/117
export const getCurrentUserMemberships = ormCreateSelector(
  orm,
  state => state.orm,
  (session) =>
    session.Membership.all().toModelArray()
)

export const getAllCommunitiesSettings = createSelector(
  getCurrentUserMemberships,
  memberships => ({
    sendEmail: every(m => m.settings && m.settings.sendEmail, memberships),
    sendPushNotifications: every(m => m.settings && m.settings.sendPushNotifications, memberships)
  })
)

export const getMessageSettings = createSelector(
  getMe,
  me => me && ({
    sendEmail: includes(me.settings && me.settings.dmNotifications, ['email', 'both']),
    sendPushNotifications: includes(me.settings && me.settings.dmNotifications, ['push', 'both'])
  })
)

export function mapStateToProps (state, props) {
  const currentUser = getMe(state, props)
  const memberships = getCurrentUserMemberships(state, props)
  const blockedUsers = getBlockedUsers(state, props)
  const allCommunitiesSettings = getAllCommunitiesSettings(state, props)
  const messageSettings = getMessageSettings(state, props)

  const confirm = get('FullPageModal.confirm', state)
  const fetchPending = state.pending[FETCH_FOR_CURRENT_USER]
  const queryParams = {
    registered: getQuerystringParam('registered', null, props)
  }

  return {
    currentUser,
    memberships,
    blockedUsers,
    confirm,
    fetchPending,
    allCommunitiesSettings,
    messageSettings,
    queryParams
  }
}

export const mapDispatchToProps = {
  fetchForCurrentUser,
  updateUserSettings,
  unBlockUser,
  leaveCommunity,
  loginWithService,
  unlinkAccount,
  setConfirmBeforeClose,
  updateMembershipSettings,
  updateAllMemberships,
  registerStripeAccount
}

export function mergeProps (stateProps, dispatchProps, ownProps) {
  const { confirm } = stateProps
  const { setConfirmBeforeClose } = dispatchProps
  const setConfirm = newState => {
    if (newState === confirm) return
    return setConfirmBeforeClose(newState)
  }
  return {
    ...stateProps,
    ...dispatchProps,
    ...ownProps,
    setConfirm
  }
}

export default component => connect(mapStateToProps, mapDispatchToProps, mergeProps)(component)
