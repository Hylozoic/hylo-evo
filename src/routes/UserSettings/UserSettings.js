import React, { PropTypes, Component } from 'react'
import './UserSettings.scss'
import AccountSettings from './AccountSettings/AccountSettings'
import CommunitySettings from './CommunitySettings/CommunitySettings'
const { object, func } = PropTypes
import FullPageModal from 'routes/FullPageModal'

export default class UserSettings extends Component {
  static propTypes = {
    currentUser: object,
    onClose: func
  }

  componentDidMount () {
    this.props.fetchUserSettings()
  }

  componentDidUpdate (prevProps, prevState) {
    if (prevProps.currentUser !== this.props.currentUser) {
      this.props.fetchUserSettings()
    }
  }

  render () {
    const {
      currentUser,
      communities,
      updateUserSettings,
      leaveCommunity,
      loginWithService
    } = this.props

    return <FullPageModal
      content={[
        {
          name: 'Account',
          path: '/settings',
          component: <AccountSettings
            currentUser={currentUser}
            updateUserSettings={updateUserSettings}
            loginWithService={loginWithService} />
        },
        {
          name: 'Communities',
          path: '/settings/communities',
          component: <CommunitySettings
            communities={communities}
            leaveCommunity={leaveCommunity} />
        }
      ]} />
  }
}
