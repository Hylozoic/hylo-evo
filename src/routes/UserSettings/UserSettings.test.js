import UserSettings from './UserSettings'
import { shallow } from 'enzyme'
import React from 'react'

describe('UserSettings', () => {
  it('renders correctly', () => {
    const currentUser = {id: 1}
    const memberships = [{id: 2}, {id: 3}]
    const updateUserSettings = () => {}
    const leaveCommunity = () => {}
    const loginWithService = () => {}
    const unlinkAccount = () => {}
    const setConfirm = () => {}
    const updateMembershipSettings = () => {}

    const wrapper = shallow(<UserSettings
      currentUser={currentUser}
      memberships={memberships}
      updateUserSettings={updateUserSettings}
      leaveCommunity={leaveCommunity}
      loginWithService={loginWithService}
      unlinkAccount={unlinkAccount}
      setConfirm={setConfirm}
      updateMembershipSettings={updateMembershipSettings} />)
    expect(wrapper).toMatchSnapshot()
  })
})
