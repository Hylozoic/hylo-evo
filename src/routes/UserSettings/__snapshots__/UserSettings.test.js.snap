// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserSettings renders correctly 1`] = `
<withRouter(Connect(FullPageModal))
  content={
    Array [
      Object {
        "component": <AccountSettingsTab
          currentUser={
            Object {
              "id": 1,
            }
          }
          fetchPending={true}
          loginWithService={[Function]}
          setConfirm={[Function]}
          unlinkAccount={[Function]}
          updateUserSettings={[Function]}
        />,
        "name": "Account",
        "path": "/settings",
      },
      Object {
        "component": <CommunitySettingsTab
          leaveCommunity={[Function]}
          memberships={
            Array [
              Object {
                "id": 2,
              },
              Object {
                "id": 3,
              },
            ]
          }
        />,
        "name": "Communities",
        "path": "/settings/communities",
      },
      Object {
        "component": <NotificationSettingsTab
          allCommunitiesSettings={undefined}
          currentUser={
            Object {
              "id": 1,
            }
          }
          memberships={
            Array [
              Object {
                "id": 2,
              },
              Object {
                "id": 3,
              },
            ]
          }
          messageSettings={undefined}
          updateAllMemberships={undefined}
          updateMembershipSettings={[Function]}
          updateUserSettings={[Function]}
        />,
        "name": "Notifications",
        "path": "/settings/notifications",
      },
      Object {
        "component": <BlockedUsersTab
          blockedUsers={undefined}
          loading={true}
          unBlockUser={undefined}
        />,
        "name": "Blocked Users",
        "path": "/settings/blocked-users",
      },
      Object {
        "component": <PasswordSettingsTab
          currentUser={
            Object {
              "id": 1,
            }
          }
          setConfirm={[Function]}
          updateUserSettings={[Function]}
        />,
        "name": "Password",
        "path": "/settings/password",
      },
    ]
  }
/>
`;
